source/sigv4.c

```diff
@@ -2741,27 +2787,36 @@ static SigV4Status_t generateSigningKey( const SigV4Parameters_t * pSigV4Params,
     SigV4Status_t returnStatus = SigV4Success;
     int32_t hmacStatus = 0;
     char * pSigningKeyStart = NULL;
+    bool isBufferSpaceSufficient = true;
 
     assert( pSigV4Params != NULL );
     assert( pHmacContext != NULL );
     assert( pSigningKey != NULL );
     assert( pBytesRemaining != NULL );
 
-    hmacStatus = hmacKey( pHmacContext,
-                          SIGV4_HMAC_SIGNING_KEY_PREFIX,
-                          SIGV4_HMAC_SIGNING_KEY_PREFIX_LEN );
-
     /* To calculate the final signing key, this function needs at least enough
      * buffer to hold the length of two digests since one digest is used to
      * calculate the other. */
     if( *pBytesRemaining < ( pSigV4Params->pCryptoInterface->hashDigestLen * 2U ) )
     {
-        returnStatus = SigV4InsufficientMemory;
+        isBufferSpaceSufficient = false;
         LOG_INSUFFICIENT_MEMORY_ERROR( "generate signing key",
                                        ( pSigV4Params->pCryptoInterface->hashDigestLen * 2U ) - *pBytesRemaining );
     }
 
-    if( hmacStatus == 0 )
+    if( isBufferSpaceSufficient )
+    {
+        /* Fill the key prefix, "AWS4" in the HMAC context cache.
+         * The prefix is part of the key for the first round of HMAC operation:
+         * HMAC("AWS4" + SecretKey, Date) */
+        hmacStatus = hmacAddKey( pHmacContext,
+                                 SIGV4_HMAC_SIGNING_KEY_PREFIX,
+                                 SIGV4_HMAC_SIGNING_KEY_PREFIX_LEN,
+                                 true /* Is key prefix. */ );
+        assert( hmacStatus == 0 );
```

Comment:
Why assert this? That may certainly not be the case depending on how the hash functions are implemented.

Reply:
Adding the prefix does not involve hashing operations but only involves adding the prefix string to the cache. Thus, I added the assert.